### PR TITLE
Fix to output checked checkboxes properly as concatenated string

### PR DESCRIPTION
**Pull request by martynhoyer**
Just something I noticed while setting this up in a project; if you have a fieldset of checkboxes in your form and the user checks multiples, the data that gets submitted only contains the last checked checkbox value.

The fix in this PR creates a string and concatenates the values (with a comma and a space) of each `checked` checkbox within the loop and outputs it at the end (removing the last comma/space along the way).

This seems to give the desired result, but a code review/second pair of eyes would be appreciated!

### CODE REVIEW

File: form-submission-handler.js [hunk 1]
```diff
@@ -19,13 +19,16 @@ function getFormData() {
   var data = {};
   fields.forEach(function(k){
     data[k] = elements[k].value;
+    var str = "";
     if(elements[k].type === "checkbox"){
-      data[k] = elements[k].checked;
+      str = str + elements[k].checked + ", ";
+      data[k] = str.slice(0, -2);
```
Comment (nelsonic) [inline form-submission-handler.js]: might be useful to have an inline comment here explaining why we want `str.slice(0, -2)` ... 🤔

Comment (martynhoyer) [inline form-submission-handler.js]: Good call, will do.

**Comment by mckennapsean** [comment]
`elements[k]` is referring to a single form item, isn't it?

There is a loop for every element `k`, but you reset `str=""` each time. So I don't see how this could work (but I may be mistaken here!). Have you tested it?

For the form use-case, you have a single form "name" with multiple checkboxes in it, yes? I would argue that you would want an array of values (but I understand the data may not look the same, so a string may suffice). I would have to play with it to see.

**Comment by martynhoyer** [comment]
I have tested it, yes.

The use case is a single field, with multiple checkboxes in it, so something like:

```
Interested in? 

- [ ] HTML
- [ ] JS
- [ ] CSS
- [ ] React

```

Currently if a user selects all four of those items, the Google spreadsheet cell and the email that gets sent only reports the last `checked` item. With the proposed method, we build up a single blob of data (the string `str`) each time the loop finds a "checked" item in a field (otherwise we'd have to put each item in a different cell in the spreadsheet which is a bit of a headache. 
The scope of `str` is inside the `forEach` so it's reset for each field, but not globally.

**Comment by mckennapsean** [comment]
Yes, I can agree that there is a bug with the way the form currently works, and you got that right!

Check out `test.html`, which just got merged into the master branch. I created this to help test different form combinations. There is one combination which has multiple checkboxes (`name="checkboxes"`), and I reproduced the error you mentioned with this test.

However, when I pulled in your code, it doesn't quite work, at least in Google Chrome. What browser did you test in? I know that Edge (and I think Chrome) treat a collection of radio buttons and checkboxes as a special item, which is an array that has a length. So, I think that the code you added never gets called, at least in my browser. I haven't stepped through to double-check, but some quick testing hints that is the issue. Does that make sense?

**Comment by mckennapsean** [comment]
Apologies, deleted my old comment since I had messed up testing your code! I can agree it works for Chrome (but we should test it in other browsers still).

Check out `test.html`, which just got merged into the master branch. I created this to help test different form combinations. There is one combination which has multiple checkboxes (`name="checkboxes"`), and I reproduced the error you mentioned with this test. Your fix did solve the issue in my browser.

However, we have to be careful with some of this code. For example, in Google Chrome, and Edge, it treats a collection of radio buttons or checkboxes as an array. Is this true in other browsers? I am not sure, but I thought behavior varied a bit when I added this code in. This is why I think we should test, at least in Firefox & Safari before we merge. Does this make sense? Thanks for the commit & PR!

**Comment by martynhoyer** [comment]
Yes, that makes sense. I've tested in latest Chrome, Firefox and Edge (via BrowserStack) after seeing the Edge-specific stuff in there, but I'll happily do some more thorough testing when I get around to adding those extra comments.

**Comment by mckennapsean** [comment]
No worries! I just did some tests myself, and it all checked out. Nice fix! With the comments it'll be a nice addition, let us know when you are done!

And, it may help to go through and clean up that script's code sometime, it works but could stand more comments and better variable names here and there. :) But that's a separate issue.

Nice job, @martynhoyer !

**Comment by martynhoyer** [comment]
Comments added!

**Comment by mckennapsean** [comment]
Thanks @martynhoyer ! Looks good to me.

**Comment by JeremiahMartinez** [comment]
Hi, I created an account just for this issue.  I tried using martynhoyer's fix but my spreadsheet is still only showing the first box checked, not concatenating them as a string.  The objective of this code is for the user to select checkboxes for email recipients.  I'd like the values to be strung together seperated by a semicolon and space, then used as the To: in the MailApp.sendEmail.
This is how I have my form written:
` <form id="gform" method="POST" class="pure-form pure-form-stacked"
  action="https://script.google.com/macros/s/AKfycbyvPwb-PV-XigKRMQj9k0YD1w6KaMNT2GuvYbgCO9lKGU3_bgDv/exec">

    <fieldset>
    <label for="personal" class="pure-checkbox">
        <input id="personal" type="checkbox" name="email[]" value="my_email@private_account.com">
        Gmail
    </label>
    
    <label for="business" class="pure-checkbox">
        <input id="work" type="checkbox" name="email[]" value="my_email@work_account.com">
        Business
    </label>
    </fieldset>
     <fieldset class="pure-group">
      <label for="message">Comment: </label>
      <textarea id="message" name="message" rows="10"
      placeholder="(Optional)"></textarea>
    </fieldset>

     <button class="button-success pure-button button-xlarge">
      <i class="fa fa-paper-plane"></i>&nbsp;Send</button>

  </form>`

And the function getFormData() is written:
`function getFormData() {
  var form = document.getElementById("gform");
  var elements = form.elements; // all form elements
  var fields = Object.keys(elements).map(function(k) {
    if(elements[k].name !== undefined) {
      return elements[k].name;
    // special case for Edge's html collection
    }else if(elements[k].length > 0){
      return elements[k].item(0).name;
    }
  }).filter(function(item, pos, self) {
    return self.indexOf(item) == pos && item;
  });
  var data = {};
   fields.forEach(function(k){
     data[k] = elements[k].value;
    var str = "";
     if(elements[k].type === "checkbox"){
      //data[k] = elements[k].checked;  //Apparently this is supposed to be removed
      str = str + elements[k].checked + "; ";
      data[k] = str.slice(0, -2);
    }else if(elements[k].length){
       for(var i = 0; i < elements[k].length; i++){
         if(elements[k].item(i).checked){
          //data[k] = elements[k].item(i).value;   //Apparently this is also supposed to be removed
          str = str + elements[k].item(i).value + ", ";
          data[k] = str.slice(0, -2);
         }
       }
  }});

  // add form-specific values into the data
  data.formDataNameOrder = JSON.stringify(fields);
  data.formGoogleSheetName = form.dataset.sheet || "responses"; // default sheet name
  data.formGoogleSendEmail = form.dataset.email || ""; // no email by default

  console.log(data);
  return data;
}`

For now, I've set my own email as the TO_ADDRESS just to monitor string results.  In that email, I do see both checked items as a string seperated by a comma.  The email looks like this:

> email[]
my_email@private_account.com,my_email@work_account.com
message
This is a test.

Any suggestions?  Thank you!!

**Comment by mckennapsean** [comment]
It looks like your email has both of the boxes checked, not just the first one? This PR was closed a while back, and we have a working example in the repo, along with the example html that looks close to yours: https://github.com/dwyl/html-form-send-email-via-google-script-without-server/blob/master/test.html#L85

If I recall, there may be a bug around multiple groups like checkboxes in some browsers, possibly IE.

Anyways, if you are still having an issue, I would suggest creating a new GitHub issue for it, as this page is for discussing code that was already pushed to the repo (and may or may not be relevant to your issue currently).

**Comment by JeremiahMartinez** [comment]
Thank you for the direction,  For future readers, I started a new issue here

https://github.com/dwyl/html-form-send-email-via-google-script-without-server/issues/192

**Pull request merged**